### PR TITLE
expression: Fix incorrect behavior of greatest/least when one argument's type is duration while others' types are none temporary(#30264)

### LINKED ISSUE #30264: GREATEST function with plus operations to time/date type gives wrong result.
### 1. Minimal reproduce step 

```sql
SELECT GREATEST(time '20:00:00', 120000) + 0.00;
SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0;
SELECT GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00;
```

### 2. What did you expect to see? 
```sql
mysql> SELECT GREATEST(time '20:00:00', 120000) + 0.00;
+------------------------------------------+
| GREATEST(time '20:00:00', 120000) + 0.00 |
+------------------------------------------+
|                                200000.00 |
+------------------------------------------+
1 row in set (0.00 sec)

mysql> SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0;
+---------------------------------------------------------------+
| GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0 |
+---------------------------------------------------------------+
|                                                      20050505 |
+---------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> SELECT GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00;
+------------------------------------------------------------------+
| GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00 |
+------------------------------------------------------------------+
|                                                      20050505.00 |
+------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead 
```sql

TiDB> SELECT GREATEST(time '20:00:00', 120000) + 0.00;
+------------------------------------------+
| GREATEST(time '20:00:00', 120000) + 0.00 |
+------------------------------------------+
|                                       20 |
+------------------------------------------+
1 row in set, 1 warning (0.00 sec)

TiDB> SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0;
+---------------------------------------------------------------+
| GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0 |
+---------------------------------------------------------------+
|                                                          2005 |
+---------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

TiDB> SELECT GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00;
+------------------------------------------------------------------+
| GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00 |
+------------------------------------------------------------------+
|                                                             2005 |
+------------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

### 4. What is your TiDB version? (Required)

```sql
tidb_version(): Release Version: v5.2.2
Edition: Community
Git Commit Hash: da1c21fd45a4ea5900ac16d2f4a248143f378d18
Git Branch: heads/refs/tags/v5.2.2
UTC Build Time: 2021-10-20 06:03:29
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: close #30264 

Problem Summary:

### What is changed and how it works?
It changed the comparison type of greatest/least function when one argument's type is duration and others are not. It changed to align the behavior with MySQL.
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by yibin87** [comment]
@guo-shaoge , please take a look at the fix.

File: expression/builtin_compare_test.go [hunk 1]
```diff
@@ -351,6 +351,10 @@ func TestGreatestLeastFunc(t *testing.T) {
 			[]interface{}{905969664.0, 4556, "1990-06-16 17:22:56.005534"},
 			"905969664", "1990-06-16 17:22:56.005534", false, false,
 		},
+		{
+			[]interface{}{105969664.0, 120000, types.Duration{Duration: 20*time.Hour + 0*time.Minute + 0*time.Second}},
+			"20:00:00", "105969664", false, false,
+		},
```
Comment (guo-shaoge) [inline expression/builtin_compare_test.go:357]: Maybe more case:
1. greatest(date_val, int_val)
2. greatest(year_val, int_val)

Comment (yibin87) [reply expression/builtin_compare_test.go:357]: Make Sense, I'll add more cases.

**Comment by yibin87** [comment]
\run-unit-test

**Comment by XuHuaiyu** [comment]
Test failed
``` go
[2021-12-23T07:43:58.472Z] --- FAIL: TestGreatestTimeType (0.39s)
[2021-12-23T07:43:58.472Z]     result.go:50: 
[2021-12-23T07:43:58.472Z]         	Error Trace:	result.go:50
[2021-12-23T07:43:58.472Z]         	            				integration_test.go:1574
[2021-12-23T07:43:58.472Z]         	Error:      	Not equal: 
[2021-12-23T07:43:58.472Z]         	            	expected: "[2021-10-11]\n"
[2021-12-23T07:43:58.472Z]         	            	actual  : "[2021-10-11 00:00:00]\n"
[2021-12-23T07:43:58.472Z]         	            	
[2021-12-23T07:43:58.472Z]         	            	Diff:
[2021-12-23T07:43:58.472Z]         	            	--- Expected
[2021-12-23T07:43:58.472Z]         	            	+++ Actual
[2021-12-23T07:43:58.472Z]         	            	@@ -1,2 +1,2 @@
[2021-12-23T07:43:58.472Z]         	            	-[2021-10-11]
[2021-12-23T07:43:58.472Z]         	            	+[2021-10-11 00:00:00]
[2021-12-23T07:43:58.472Z]         	            	 
[2021-12-23T07:43:58.472Z]         	Test:       	TestGreatestTimeType
[2021-12-23T07:43:58.472Z]         	Messages:   	sql:select greatest(c_d, c_d) from t1;, args:[]
```

**Pull request closed without merging**